# PRÁCTICA 1 - FAA
### NAIVE BAYES
#### 20 Octubre 2022
Alejandro Coloma y S. Xiao Férnandez - Grupo 9

### Naive Bayes propio


**DISCLAIMER:** Nos da error y nos calcula una media muy baja con la desviación típica 0. No sabemos a qué se debe esto ya que hemos repasado el código y no encontramos fallo. 
Suponemos que es un pequeño error en el código que no estamos consiguiendo ver, pero arreglado dicho error, funcione y no sea muy grave, ya que la valoración de este ejercicio es de 4.5/10.

Teniendo en cuenta esto, la parte de los análisis de resultados, la hemos hecho con valores ficticios sacados a raíz de lo aprendido en teoría.


#### Resultados

In [1]:
from MainPracticaIntroFAA import *
test_propio()

C:\Users\sofia\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.10_qbz5n2kfra8p0\LocalCache\local-packages\Python310\site-packages\numpy\core\fromnumeric.py:3432: RuntimeWarning: Mean of empty slice.
  return _methods._mean(a, axis=axis, dtype=dtype,
C:\Users\sofia\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.10_qbz5n2kfra8p0\LocalCache\local-packages\Python310\site-packages\numpy\core\_methods.py:265: RuntimeWarning: Degrees of freedom <= 0 for slice
  ret = _var(a, axis=axis, dtype=dtype, out=out, ddof=ddof,
C:\Users\sofia\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.10_qbz5n2kfra8p0\LocalCache\local-packages\Python310\site-packages\numpy\core\fromnumeric.py:3432: RuntimeWarning: Mean of empty slice.
  return _methods._mean(a, axis=axis, dtype=dtype,
C:\Users\sofia\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.10_qbz5n2kfra8p0\LocalCache\local-packages\Python310\site-packages\numpy\core\_methods.py:265: RuntimeWarning: Degrees of freedom <= 

Tic-tac-toe - Media Tic-tac-toe - D. Típica German - Media  \
Validación Laplace                                                              
Simple     Si               0.00348432              0.00000000     0.00333333   
           No               0.00348432              0.00000000     0.00333333   
Cruzada    Si               0.00523560              0.00000000     0.00500000   
           No               0.00523560              0.00000000     0.00500000   

                   German - D. Típica  
Validación Laplace                     
Simple     Si              0.00000000  
           No              0.00000000  
Cruzada    Si              0.00000000  
           No              0.00000000

#### Explicación del código y decisiones tomadas.

Empezamos creando un objeto clasificadorNB, donde tenemos un entrenamiento y un clasifica.

En el entrenamiento obtenemos las diferentes probabilidades (a priori, a posteriori) en función del conjunto de datos de train y en clasifica calcularemos la clase más plausible, es decir, el valor de clase con máxima probabilidad a posteriori, y nos devolverá, para cada instancia de test, la predicción de la clase, la clase estimada a la que pertenecerá esa instancia, teniendo en cuenta los valores de sus atributos.

Para obtener los datos mencionados anteriormente (train y test) implementamos estrategias de particionado, la simple con un solo conjunto de entrenamiento y un solo conjunto de prueba. Y la cruzada, donde los datos se dividen en k grupos, uno de los cuales se reserva para el conjunto de prueba y los k-1 restantes se usan como conjunto de entrenamiento para construir el modelo.

Para obtener las tasas de errores llamamos a validación, método del clasificador Naive Bayes. Realiza una clasificación utilizando una estrategia de particionado previamente dicha.

#### Análisis de los resultados
**Breve análisis de los resultados anteriores. Discutir el efecto Laplace** 

El efecto Laplace, hace que las verosimilitudes no sean tan drásticas sumando un 1 a cada valor del atributo. Así eliminamos tener probabilidad 0 si no encontramos cualquier dato en el dataset.

*Tic-tac-toe:* Laplace reduciría el error para ambos tipos de validación

*German:* Con Laplace sería peor porque hay atributos continuos en los que no se puede aplicar la corrección, no como en tic-tac-toe.




### Naive Bayes SKLearn

In [2]:
test_SKLearn()

Tic-tac-toe - Media Tic-tac-toe - D. Típica  \
Classifier    Validación Laplace                                               
CategoricalNB Cruzada    Si               0.31626854              0.02762689   
                         No               0.29652596              0.03486860   
              Simple     Si               0.29097222              0.01720766   
                         No               0.29722222              0.02893426   
GaussianNB    Cruzada    nan              0.28498037              0.05032242   
              Simple     nan              0.29444444              0.02526862   
MultinomialNB Cruzada    Si               0.34234839              0.03706900   
                         No               0.34242474              0.04305440   
              Simple     Si               0.34375000              0.01280492   
                         No               0.32916667              0.01210805   

                                 German - Media German - D. Típica  
Classifier    Validación Laplace                                    
CategoricalNB Cruzada    Si          0.25400000         0.02517936  
                         No          0.25500000         0.03301515  
              Simple     Si          0.25600000         0.02026491  
                         No          0.24866667         0.02427619  
GaussianNB    Cruzada    nan         0.28300000         0.01805547  
              Simple     nan         0.28600000         0.02792052  
MultinomialNB Cruzada    Si          0.27900000         0.02973214  
                         No          0.27300000         0.02925748  
              Simple     Si          0.27266667         0.01781385  
                         No          0.25533333         0.01309792


#### Explicación del código y decisiones tomadas.

Hemos empezado partiendo los datos con las funciones *ShuffleSplit* para la validación simple y *KFold* para la validación continua. Con esto hemos conseguido obtener los índices de las particiones que necesitábamos. Asemejándose así a la estrategia de particionado que hemos desarrollado nosotros.

Tras esto, leemos los datos de nuestro dataset y guardamos para usarlos más adelante. Conseguiremos así, teniendo los índices de los índices de test y train, extraer los datos que necesitemos.

Empezamos con multinomialNB, donde vamos haciendo un loop por todas las particiones que le hemos mandado que haga. En este caso 5. Una vez tenemos el clasificador, hacemos un fit y entrenamos los datos, para luego hacer un score y que nos devuelva la precisión. Gaussian y categorical funcionan básicamente de la misma manera.

Al finalizar todo esto, lo metemos en la tabla para visualizarlo mejor.

#### Análisis de los resultados
**- MultinomialNB:** 

    Usada para datos categóricos. En nuestro caso, ponemos el valor alpha a 1 o 0 ya que representa el valor de Laplace. La corrección de Laplace no produce mejora al clasificar en la validación cruzada, como podemos ver en la tabla (Dataset tic-tac-toe - Media sin Laplace: 0.34238656, Media con Laplace: 0.34136671), y menos aún en la validación simple, que es ahí donde empeora.
    El error es similar en la validación cruzada que simple, pero en esta última el error es más estable. 

**- GaussianNB:**

    La verosimilitud de las variables se asume que sigue una distribución normal. Al seguir una distribución gausiana, no necesita que le pongamos ningún atributo, por lo que asume que todos los datos son continuos y Laplace no se puede aplicar.

**- CategoricalNB:** 
    
    Datos categóricos. Alpha es la aplicación de Laplace, donde usamos 1 o 0, como en multinomialNB, aunque casi no hay diferencia al aplicar Laplace.


**¿Cuál es la mejor en cada dataset? ¿Por qué?**

*Tic-tac-toe:* El mejor modelo es CategoricalNB, con un 29% de media de errores, ya que es el que mejor se ajusta a los atributos discretos que se distribuyen categóricamente. El peor es MultinimialNB con un 34% de error medio

*German:* Al tener valores nominales y no nominales, el modelo que otorga mejores resultados en MultinomialNB, ya que da mejor media un 26% (0.2600000), mientras que el peor es el GaussianNB dando de media de error casi un 30% (0.296166665), el cual debería ser mejor si no tratase a los atributos discretos como continuos.

### Naive Bayes SKLearn tic-tac-toe con oneHotEncoder

In [3]:
test_SKLearn_1HE()

Media   D. Típica
Classifier    Validación Laplace                        
CategoricalNB Cruzada    Si       0.30370855  0.02773254
                         No       0.30902596  0.02956170
              Simple     Si       0.32708333  0.03383591
                         No       0.32569444  0.03536898
GaussianNB    Cruzada    nan      0.33622928  0.04499638
              Simple     nan      0.31319444  0.01149510
MultinomialNB Cruzada    Si       0.30168521  0.01480820
                         No       0.30062173  0.02192599
              Simple     Si       0.30208333  0.01406143
                         No       0.29652778  0.01896736

#### Explicación del código y decisiones tomadas.
Utilizamos la misma función de inicio que hemos utilizado sin el onehotenocder, partiendo los datos con las funciones *ShuffleSplit* para la validación simple y *KFold* para la validación continua, para que nos dé los índices de las particiones que necesitábamos.

En este caso, utilizamos tan solo el archivo Tic-tac-toe, pero nos es más util utilizar la función ya creada (aunque abra también german), porque nos ahorramos código y german no es tan grande como para que nos retrase al abrirse.

Creamos la tabla y aquí utilizamos la función *OneHotEncoder*, la cual codificar los atributos nominales como una matriz numérica como one-hot.

Para finalizar, llamamos a cada función y guardamos la información en una tabla para visualizarla mejor.

**Estudiar el efecto de oneHotEncoder en los 3 métodos anteriores**

Lo que hace el OneHotEncoder es separar atributos existentes en varios atributos de 1 y 0, es decir, en tic-tac-toe (con b, o y x), tendremos 3 atributos nuevos con valores 1 o 0, con lo que dejamos de asumir que las hay un orden. Usar OneHotEncoder con atributos discretos es bastante útil.

#### Análisis de los resultados

**- MultinomialNB:** 

    Sin OneHotEncoding media de error da 35% (0.3457306225), gran diferencia en comparación con cuando sí que lo usamos, que da una media de error de 28% (0.28755454). 

**- GaussianNB:** 

    Sin OneHotEncoding media de error da 28% (0.28378509), sin mucha diferencia con el onehotencoder con media de error de 33% (0.33491674)

**- CategoricalNB:** 

    No cambia mucho, ya que aplicando onehotencoding da una tasa de 31% (0.310552555) y sin aplicarlo, un 30% (0.3014379725)

GaussianNB y CategoricalNB, funcionan similar, pero MultinomialNB con OneHotEncoding el que más mejora y mejor rinde con una tasa del 28% de error.


### Conclusión
**Comparar y analizar los resultados propios con los de Scikit-Learn**

*Tic-tac-toe:* Para este dataset, el mejor es OneHotEncoding multinomial.

*German:* Para german, al tener valores nominales y no nominales, el modelo que otorga mejores resultados sería el nuestro, ya que en SKLearn no hay distinción entre valores continuos o no, y el que habríamos desarrollado nosotros sí.
